File: app/src/main/java/com/mtime/player/receivers/PauseAdCover.java
package com.mtime.player.receivers;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.kk.taurus.playerbase.event.EventKey;
import com.kk.taurus.playerbase.event.OnPlayerEventListener;
import com.kk.taurus.playerbase.receiver.BaseCover;
import com.kk.taurus.playerbase.receiver.IReceiverGroup;
import com.mtime.R;
import com.mtime.applink.ApplinkManager;
import com.mtime.base.imageload.ImageHelper;
import com.mtime.base.imageload.ImageProxyUrl;
import com.mtime.bussiness.video.bean.VideoAdBean;
import com.mtime.player.AdHelper;
import com.mtime.player.DataInter;
import com.mtime.player.PlayerHelper;
import com.mtime.player.bean.VideoInfo;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by mtime on 2017/10/18.
 * 暂停广告控制层
 */

public class PauseAdCover extends BaseCover {

    public static final String KEY = "pause_ad_cover";

    private final String TAG = "PauseAdCover";

    private TextView mAdTag;
    private ImageView mAdImage;
    private ImageView mCloseIcon;
    private RelativeLayout mAdContainer;

    private boolean isAdShow;
    private boolean isFullScreen;

    private int mIndex;

    private int mPauseCount;
    private boolean needLargeDelay;

    private final float rate_w = 0.6f;
    private final float rate_h = 0.61f;

    private int image_w;
    private int image_h;

    private List<VideoAdBean.AdItem> mAdItems = new ArrayList<>();

    private final int MSG_CODE_SHOW_IMAGE = 333;
    private boolean isEditDanmu;

    private final Handler mHandler = new Handler(Looper.getMainLooper()){
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            switch (msg.what){
                case MSG_CODE_SHOW_IMAGE:
                    final VideoAdBean.AdItem item = (VideoAdBean.AdItem) msg.obj;
                    Log.d(TAG,"id = " + item.getaID() + " url = " + item.getImage());
                    //display image
                    ImageHelper.with(getContext(), ImageProxyUrl.SizeType.CUSTOM_SIZE, ImageProxyUrl.ClipType.FIX_WIDTH_AND_HEIGHT)
                            .override(image_w, image_h)
                            .load(item.getImage())
                            .view(mAdImage)
                            .error(R.drawable.collections_default)
                            .placeholder(R.drawable.collections_default)
                            .showload();
                    final Bundle bundle = new Bundle();
                    bundle.putInt(EventKey.INT_ARG1,item.getaID());
                    bundle.putInt(EventKey.INT_ARG2,mIndex+1);
                    //notify show event for statistics
                    notifyReceiverEvent(DataInter.ReceiverEvent.EVENT_CODE_ON_PAUSE_AD_SHOW,bundle);
                    //handle applink for ad image
                    final String appLink = item.getApplinkData();
                    Log.d(TAG,"applink = " + appLink);
                    mAdImage.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            //notify ad click event for statistics
                            notifyReceiverEvent(DataInter.ReceiverEvent.EVENT_CODE_ON_PAUSE_AD_CLICK, bundle);
                            ApplinkManager.jump(getContext(), appLink, getPageRefer());
                        }
                    });

                    //handle ad tag
                    mAdTag.setVisibility(View.GONE);
                    if(item.isShowTag()){
                        Log.d(TAG,"showTag ..." + item.getTagDesc());
                        mAdTag.setVisibility(View.VISIBLE);
                        mAdTag.setText(item.getTagDesc());
                    }

                    //delay show next ad item
                    mIndex++;
                    int size = mAdItems.size();
                    mIndex = mIndex%size;
                    Message message = Message.obtain();
                    message.what = MSG_CODE_SHOW_IMAGE;
                    message.obj = mAdItems.get(mIndex);

                    //default use item delay duration
                    long delayTime = item.getDuration() * 1000;
                    if(mPauseCount > 0 && needLargeDelay){
                        needLargeDelay = false;
                        delayTime = delayTime * 3;
                    }

                    //loop next item
                    mHandler.sendMessageDelayed(message, delayTime);
                    break;
            }
        }
    };

    public PauseAdCover(Context context) {
        super(context);
    }

    private String getPageRefer(){
        return getGroupValue().getString(DataInter.Key.KEY_STATISTICS_PAGE_REFER);
    }

    @Override
    public void onReceiverBind() {
        super.onReceiverBind();
        mAdTag = findViewById(R.id.layout_pause_ad_cover_tag_tv);
        mAdContainer = findViewById(R.id.layout_pause_ad_cover_container_rl);
        mAdImage = findViewById(R.id.iv_ad_img);
        mCloseIcon = findViewById(R.id.iv_close_ad);
        mCloseIcon.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(isAdShow){
                    //hidden ad
                    setAdState(false);
                }
            }
        });

        getGroupValue().registerOnGroupValueUpdateListener(onGroupValueUpdateListener);
    }

    private final IReceiverGroup.OnGroupValueUpdateListener onGroupValueUpdateListener =
            new IReceiverGroup.OnGroupValueUpdateListener() {
        @Override
        public String[] filterKeys() {
            return new String[]{
                    DataInter.Key.KEY_IS_FULL_SCREEN,
                    DataInter.Key.KEY_IS_LANDSCAPE_EDIT_DANMU
            };
        }

        @Override
        public void onValueUpdate(String s, Object o) {
            if(DataInter.Key.KEY_IS_FULL_SCREEN.equals(s)){
                isFullScreen = (boolean) o;
                setAdShowMode(isFullScreen);
            }else if(DataInter.Key.KEY_IS_LANDSCAPE_EDIT_DANMU.equals(s)){
                isEditDanmu = (boolean) o;
            }
        }
    };

    /**
     * 大小屏切换时，动态调整广告图的大小
     * @param isFullScreen
     */
    private void setAdShowMode(boolean isFullScreen){
        int screen_w = PlayerHelper.getScreenW(getContext());
        int screen_h = PlayerHelper.getScreenH(getContext());
        ViewGroup.LayoutParams layoutParams = mAdContainer.getLayoutParams();
        if(isFullScreen){
            int w = Math.max(screen_w, screen_h);
            int h = Math.min(screen_w, screen_h);
            image_w = (int) (w * rate_w);
            image_h = (int) (h * rate_h);
        }else{
            image_w = (int) (screen_w * rate_w);
            image_h = (int) ((screen_w * 9/16) * rate_h);
        }
        layoutParams.width = image_w;
        layoutParams.height = image_h;
        mAdContainer.setLayoutParams(layoutParams);
    }

    public void setAdState(boolean state){
        if(!state){
            mHandler.removeMessages(MSG_CODE_SHOW_IMAGE);
        }
        if(state && !isAdDataAvailable()){
            return;
        }
        if(state && isEditDanmu){
            setCoverVisibility(View.GONE);
            return;
        }
        isAdShow = state;
        if(isAdShow){
            handleDisPlayImage();
        }
        setAdShowMode(isFullScreen);
        setCoverVisibility(state?View.VISIBLE:View.GONE);
        if(!state){
            mAdImage.setImageResource(R.drawable.collections_default);
        }
    }

    /**
     * 需求：广告内容为2以上时，每次暂停时按照权重顺序轮换展示,当停留时间过长时（界定条件：后台配置的 多张轮播时间*3），每则广告循环播放3次，继续播放下一则广告，依次循环

     假设：

     广告A  权重：1  多张轮播时间：10s

     广告B  权重：2  多张轮播时间：15s

     广告C  权重：3  多张轮播时间：20s

     1、视频第一次暂停，显示广告A，当停留时间超过10s*3，则显示广告B

     2、视频继续播放，第二次暂停，停留时间超过15s*3，则显示广告C

     3、视频继续播放，第三次暂停，停留时间超过20s*3，则显示广告A
     */
    private void handleDisPlayImage() {
        if(isAdDataAvailable()){
            mIndex = 0;
            VideoAdBean.AdItem adItem = mAdItems.get(mIndex);
            int size = mAdItems.size();
            if(size>=2){
                int tempIndex = mPauseCount%size;
                adItem = mAdItems.get(tempIndex);
                mIndex = tempIndex;
                mPauseCount++;
                needLargeDelay = true;
            }else{
                mPauseCount = 0;
            }
            Message message = Message.obtain();
            message.what = MSG_CODE_SHOW_IMAGE;
            message.obj = adItem;
            mHandler.sendMessage(message);
        }
    }

    @Override
    public void onPlayerEvent(int eventCode, Bundle bundle) {
        switch (eventCode){
            case OnPlayerEventListener.PLAYER_EVENT_ON_RESUME:
                setAdState(false);
                break;

            case OnPlayerEventListener.PLAYER_EVENT_ON_PAUSE:
                if(!isDanmuCoverInEditState()){
                    setAdState(true);
                }
                break;
            case OnPlayerEventListener.PLAYER_EVENT_ON_PROVIDER_DATA_START:
            case OnPlayerEventListener.PLAYER_EVENT_ON_DATA_SOURCE_SET:
                if(isAdShow){
                    setAdState(false);
                }
                break;
        }
    }

    private boolean isDanmuCoverInEditState(){
        return getGroupValue().getBoolean(DataInter.Key.KEY_DANMU_COVER_IN_LANDSCAPE_EDIT_STATE);
    }

    @Override
    public void onErrorEvent(int i, Bundle bundle) {

    }

    private VideoInfo getCurrentVideoInfo(){
        return getGroupValue().get(DataInter.Key.KEY_VIDEO_INFO);
    }

    private boolean isAdDataAvailable(){
        VideoInfo videoInfo = getCurrentVideoInfo();
        if(videoInfo==null)
            return false;
        List<VideoAdBean.AdItem> adList = AdHelper.getAdList(
                String.valueOf(videoInfo.getVid()), Integer.parseInt(videoInfo.getSourceType()));
        if(adList==null || adList.size()<=0)
            return false;
        mAdItems = adList;
        return true;
    }

    @Override
    public void onReceiverEvent(int eventCode, Bundle bundle) {

    }

    @Override
    public View onCreateCoverView(Context context) {
        return View.inflate(context, R.layout.layout_pause_ad_cover,null);
    }

    @Override
    public void onReceiverUnBind() {
        super.onReceiverUnBind();
        mHandler.removeMessages(MSG_CODE_SHOW_IMAGE);
        getGroupValue().unregisterOnGroupValueUpdateListener(onGroupValueUpdateListener);
    }

    @Override
    public int getCoverLevel() {
            return levelLow(DataInter.CoverLevel.COVER_LEVEL_PAUSE_AD);
    }
}

File: player/src/main/java/com/kotlin/android/player/receivers/PauseAdCover.kt
package com.kotlin.android.player.receivers

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import com.kk.taurus.playerbase.event.EventKey
import com.kk.taurus.playerbase.event.OnPlayerEventListener
import com.kk.taurus.playerbase.receiver.BaseCover
import com.kk.taurus.playerbase.receiver.IReceiverGroup.OnGroupValueUpdateListener
import com.kotlin.android.app.data.entity.player.VideoAdBean
import com.kotlin.android.app.data.entity.player.VideoInfo
import com.kotlin.android.image.coil.ext.loadImage
import com.kotlin.android.ktx.ext.core.gone
import com.kotlin.android.ktx.ext.dimension.screenHeight
import com.kotlin.android.ktx.ext.dimension.screenWidth
import com.kotlin.android.ktx.ext.core.visible
import com.kotlin.android.mtime.ktx.ext.ShapeExt
import com.kotlin.android.player.DataInter
import com.kotlin.android.player.R
import com.kotlin.android.player.ad.AdHelper
import java.util.*

/**
 * create by lushan on 2020/8/31
 * description: 暂停广告控制层
 */
class PauseAdCover(var ctx: Context) : BaseCover(ctx) {

    private val TAG = "PauseAdCover"
    private var mAdTag: TextView? = null
    private var mAdImage: ImageView? = null
    private var mCloseIcon: ImageView? = null
    private var mAdContainer: RelativeLayout? = null
    private var isAdShow = false
    private var isFullScreen = false

    private var mIndex = 0

    private var mPauseCount = 0
    private var needLargeDelay = false

    private val rate_w = 0.6f
    private val rate_h = 0.61f

    private var image_w = 0
    private var image_h = 0
    private val MSG_CODE_SHOW_IMAGE = 333
    private var isEditDanmu = false
    private var mAdItems: List<VideoAdBean.AdItem> = ArrayList<VideoAdBean.AdItem>()
    private val mHandler: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            when (msg.what) {
                MSG_CODE_SHOW_IMAGE -> {
                    val item: VideoAdBean.AdItem = msg.obj as VideoAdBean.AdItem
                    Log.d(TAG, "id = " + item.aID + " url = " + item.image)
                    //display image
                    mAdImage?.loadImage(
                        data = item.image.orEmpty(),
                        width = image_w,
                        height = image_h,
                        defaultImgRes = R.drawable.collections_default
                    )
                    val bundle = Bundle()
                    bundle.putInt(EventKey.INT_ARG1, item.aID.toInt())
                    bundle.putInt(EventKey.INT_ARG2, mIndex + 1)
                    //notify show event for statistics
                    notifyReceiverEvent(DataInter.ReceiverEvent.EVENT_CODE_ON_PAUSE_AD_SHOW, bundle)
                    //handle applink for ad image
                    val appLink: String = item.applinkData.orEmpty()
                    Log.d(TAG, "applink = $appLink")
                    mAdImage?.setOnClickListener { //notify ad click event for statistics
                        notifyReceiverEvent(DataInter.ReceiverEvent.EVENT_CODE_ON_PAUSE_AD_CLICK, bundle)
                        //todo applinkManager如何处理
//                        ApplinkManager.jump(context, appLink, getPageRefer())
                    }

                    //handle ad tag
                    mAdTag?.gone()
                    if (item.isShowTag) {
                        Log.d(TAG, "showTag ..." + item.tagDesc)
                        mAdTag?.visible()
                        mAdTag?.text = item.tagDesc.orEmpty()
                    }

                    //delay show next ad item
                    mIndex++
                    val size: Int = mAdItems.size
                    mIndex %= size
                    val message = Message.obtain()
                    message.what = MSG_CODE_SHOW_IMAGE
                    message.obj = mAdItems[mIndex]

                    //default use item delay duration
                    var delayTime: Long = item.duration * 1000.toLong()
                    if (mPauseCount > 0 && needLargeDelay) {
                        needLargeDelay = false
                        delayTime *= 3
                    }

                    //loop next item
                    sendMessageDelayed(message, delayTime)
                }
            }
        }
    }

    override fun onPlayerEvent(eventCode: Int, bundle: Bundle?) {
        when (eventCode) {
            OnPlayerEventListener.PLAYER_EVENT_ON_RESUME -> setAdState(false)
            OnPlayerEventListener.PLAYER_EVENT_ON_PAUSE -> if (!isDanmuCoverInEditState()) {
                setAdState(true)
            }
            OnPlayerEventListener.PLAYER_EVENT_ON_PROVIDER_DATA_START, OnPlayerEventListener.PLAYER_EVENT_ON_DATA_SOURCE_SET -> if (isAdShow) {
                setAdState(false)
            }
        }
    }

    private fun isDanmuCoverInEditState(): Boolean {
        return groupValue.getBoolean(DataInter.Key.KEY_DANMU_COVER_IN_LANDSCAPE_EDIT_STATE)
    }

    override fun onErrorEvent(eventCode: Int, bundle: Bundle?) {
    }

    override fun onReceiverEvent(eventCode: Int, bundle: Bundle?) {
    }

    override fun onCreateCoverView(context: Context?): View {
        return View.inflate(context, R.layout.layout_pause_ad_cover, null)
    }

    override fun onReceiverBind() {
        super.onReceiverBind()
        mAdTag = findViewById(R.id.layout_pause_ad_cover_tag_tv)
        mAdContainer = findViewById(R.id.layout_pause_ad_cover_container_rl)
        mAdImage = findViewById(R.id.iv_ad_img)
        mCloseIcon = findViewById(R.id.iv_close_ad)
        mAdTag?.apply {
            ShapeExt.setShapeCorner2Color2Stroke(this, R.color.color_66000000, 0, R.color.color_ffffff, 1, false)
        }
        mCloseIcon?.setOnClickListener(View.OnClickListener {
            if (isAdShow) {
                //hidden ad
                setAdState(false)
            }
        })

        groupValue.registerOnGroupValueUpdateListener(onGroupValueUpdateListener)
    }

    private fun getCurrentVideoInfo(): VideoInfo? {
        return groupValue.get<VideoInfo>(DataInter.Key.KEY_VIDEO_INFO)
    }

    private fun isAdDataAvailable(): Boolean {
        val videoInfo: VideoInfo = getCurrentVideoInfo() ?: return false
        val adList: List<VideoAdBean.AdItem>? = AdHelper.getAdList(videoInfo.vid, videoInfo.sourceType)
        if (adList == null || adList.isEmpty()) return false
        mAdItems = adList
        return true
    }

    private val onGroupValueUpdateListener: OnGroupValueUpdateListener = object : OnGroupValueUpdateListener {
        override fun filterKeys(): Array<String> {
            return arrayOf(
                    DataInter.Key.KEY_IS_FULL_SCREEN,
                    DataInter.Key.KEY_IS_LANDSCAPE_EDIT_DANMU
            )
        }

        override fun onValueUpdate(s: String, o: Any) {
            if (DataInter.Key.KEY_IS_FULL_SCREEN == s) {
                isFullScreen = o as Boolean
                setAdShowMode(isFullScreen)
            } else if (DataInter.Key.KEY_IS_LANDSCAPE_EDIT_DANMU == s) {
                isEditDanmu = o as Boolean
            }
        }
    }

    /**
     * 大小屏切换时，动态调整广告图的大小
     * @param isFullScreen
     */
    private fun setAdShowMode(isFullScreen: Boolean) {
        val screen_w: Int = screenWidth
        val screen_h: Int = screenHeight
        val layoutParams = mAdContainer?.layoutParams
        if (isFullScreen) {
            val w = screen_w.coerceAtLeast(screen_h)
            val h = screen_w.coerceAtMost(screen_h)
            image_w = (w * rate_w).toInt()
            image_h = (h * rate_h).toInt()
        } else {
            image_w = (screen_w * rate_w).toInt()
            image_h = (screen_w * 9 / 16 * rate_h).toInt()
        }
        layoutParams?.width = image_w
        layoutParams?.height = image_h
        mAdContainer?.layoutParams = layoutParams
    }

    fun setAdState(state: Boolean) {
        if (!state) {
            mHandler.removeMessages(MSG_CODE_SHOW_IMAGE)
        }
        if (state && !isAdDataAvailable()) {
            return
        }
        if (state && isEditDanmu) {
            setCoverVisibility(View.GONE)
            return
        }
        isAdShow = state
        if (isAdShow) {
            handleDisPlayImage()
        }
        setAdShowMode(isFullScreen)
        setCoverVisibility(if (state) View.VISIBLE else View.GONE)
        if (!state) {
            mAdImage?.setImageResource(R.drawable.collections_default)
        }
    }

    /**
     * 需求：广告内容为2以上时，每次暂停时按照权重顺序轮换展示,当停留时间过长时（界定条件：后台配置的 多张轮播时间*3），每则广告循环播放3次，继续播放下一则广告，依次循环
     *
     * 假设：
     *
     * 广告A  权重：1  多张轮播时间：10s
     *
     * 广告B  权重：2  多张轮播时间：15s
     *
     * 广告C  权重：3  多张轮播时间：20s
     *
     * 1、视频第一次暂停，显示广告A，当停留时间超过10s*3，则显示广告B
     *
     * 2、视频继续播放，第二次暂停，停留时间超过15s*3，则显示广告C
     *
     * 3、视频继续播放，第三次暂停，停留时间超过20s*3，则显示广告A
     */
    private fun handleDisPlayImage() {
        if (isAdDataAvailable()) {
            mIndex = 0
            var adItem: VideoAdBean.AdItem = mAdItems[mIndex]
            val size: Int = mAdItems.size
            if (size >= 2) {
                val tempIndex = mPauseCount % size
                adItem = mAdItems.get(tempIndex)
                mIndex = tempIndex
                mPauseCount++
                needLargeDelay = true
            } else {
                mPauseCount = 0
            }
            val message = Message.obtain()
            message.what = MSG_CODE_SHOW_IMAGE
            message.obj = adItem
            mHandler.sendMessage(message)
        }
    }

    override fun onReceiverUnBind() {
        super.onReceiverUnBind()
        mHandler.removeMessages(MSG_CODE_SHOW_IMAGE)
        groupValue.unregisterOnGroupValueUpdateListener(onGroupValueUpdateListener)
    }

    override fun getCoverLevel(): Int {
        return levelLow(DataInter.CoverLevel.COVER_LEVEL_PAUSE_AD)
    }

}